### PR TITLE
[DNM] Experimental web interface

### PR DESCRIPTION
**Pull request by jcsp**
(obviously we'll rename the module from guilolz to something more sensible... Dashboard McDashboardface?)

### CODE REVIEW

File: src/mgr/Mgr.cc [hunk 1]
```diff
@@ -418,6 +421,15 @@ void Mgr::handle_osd_map()
   daemon_state.cull(CEPH_ENTITY_TYPE_OSD, names_exist);
 }
 
+void Mgr::handle_log(MLog *m)
+{
+  for (const auto &e : m->entries) {
+    std::ostringstream ss;
+    ss << e.stamp << " " << e.who.name << " " << e.prio << " " << e.msg;
```
Comment (dmick) [inline src/mgr/Mgr.cc]: Not that you're asking for review, but, browsing this: ss is thrown away here, right?  Is this for gdb, or dead code, or?...

**Comment by liewegas** [comment]
I kinda like guilolz!

**Comment by yuyuyu101** [comment]
submodule is better?

**Comment by ktdreyer** [comment]
> submodule is better?

How is a submodule better? (There are too many submodules already IMHO)

**Comment by liewegas** [comment]
Since this isn't an external project I don't think a submodule makes sense.

**Comment by alfredodeza** [comment]
SInce this is labeled as "experimental" (proof of concept?) why the need to have in ceph core? Wouldn't it be more flexible to make it external?

The PR also has several changes that don't look related to the implementation of a gui (e.g. https://github.com/ceph/ceph/pull/13725/commits/eddd8e7a31403f609911c03e49776e69823225bc )

**Comment by alfredodeza** [comment]
A couple of other things I forgot to add:

We've been discussing on interfacing with ceph-mgr, and having an example project out there that lives out-of-source would be an excellent point of reference for others. This might be close but, would be nicer/better as something external.

For packaging and releases this also becomes a problem, we struggle a lot every time we add dependencies and files. This looks like we need to package Django, and ensure that Jinja2 and cherrypy are compatible with the runtime of the web UI. Pulling in those and making sure they work means also to maintaining tests. 

Specific Python library versions, deployment, packaging, and tests would be far easier to implement and maintain if this web UI is separate.

**Comment by jcsp** [comment]
@alfredodeza the main practical reason for putting these types of module in the Ceph tree is to saves us from maintaining a stable interface to out-of-tree stuff.  It does not necessarily mean we have to package all their dependencies etc, or that we have to commit to maintaining them in perpetuity.  The dependencies aspect would be up to the various vendors/interested parties to decide which modules they want to ship/switch on, and then only deal with dependencies for modules they care about.

The other reason for putting things like this in the tree is that there's value to having them present by default, without users having to go and engage with another project, or get some different packages.  It's increasingly looking like having a stats client, or a basic web UI is considered a basic day 1 expectation, rather than an optional addon.

BTW this only depends on jinja2 and cherrypy, there is no Django (unlike the rest module).

**Comment by alfredodeza** [comment]
@jcsp I do agree in the value of presenting something by default to fulfill the "day 1 expectation". You mention both users and vendors, but implementing this with the expectation that they need to deal with the dependency aspect will create another usability issue: the project is shipping something that might not work out of the box.

Most (if not all) distro package maintainers will not accept having undefined dependencies (what version of cherrypy and jinja2 will this work with?) or any vendored libraries (e.g. rivetjs). And it also puts another strain on us who package and distribute Ceph.

Could you also elaborate how adding these files (e.g. the CSS and JS files) save us from some maintenance? Or do you just mean the part where the UI interacts with ceph-mgr? The latter is not that complicated to do (we already do version parity for other out-of-tree projects), unless you are expecting the ceph-mgr API to be very volatile, in which case it sounds like we should wait for adding interfaces like a UI.

**Comment by jcsp** [comment]
I'm skeptical that including a static/ dir including javascript libraries would be an issue.  This is the same thing that e.g. cobbler-web or graphite-web does.  I don't think there's any reasonable expectation that every webapp on a server is going to use the same jquery.js file.

Cherrypy and jinja should already be there in centos7 and ubuntu xenial (haven't checked trusty) -- there was forethought to the pain of packaging any python deps that weren't already there.  This PR didn't update any dependency stuff but that wasn't necessarily assuming that people would have to do it by hand, I just didn't get that far.

Aside from dependencies, I think we've got several overlapping questions here:
1. Should we ship this at all?
2. Should it be in ceph-mgr package or a separate package?
3. Should it be in the ceph tree or elsewhere?

1 - an open question, this code ended up in a PR so that we could get more discussion around it.
2 - I'm indifferent to, although we would definitely want it to be installed by default (i.e. possibly a dependency of the ceph metapackage but not a hard dependency of ceph-mgr)
3 - if it's going to exist, it does need to be in a Ceph tree -- the C++<->Python interface is intentionally not a stable API for out of tree projects to consume.  It's there to enable writing code in python, not to enable interfacing with ceph from external projects.  The intention is that people who want a stable interface (e.g. a REST API) are free to maintain in-tree mgr modules that provide those interfaces.

**Comment by andrewschoen** [comment]
@jcsp Could you expand more on the decision to require code that interacts with ``ceph-mgr`` to be in-tree? I tend to agree with @alfredodeza here that a python project that interacts with ``ceph-mgr`` but lives outside of ceph.git could move quicker and would be easier to test/package. Also, I think your concern about interoperability with the ``ceph-mgr`` API would be a responsibility of this new project, not ceph.

Allowing out-of-tree projects would allow users of ceph to build their own python projects or scripts that interact with ``ceph-mgr`` more easily and without modifying ceph source. I'm thinking this might encourage more usage of ``ceph-mgr``. Perhaps I'm completely misunderstanding the purpose of ``ceph-mgr`` and it's python runtime though.

**Comment by tchaikov** [comment]
maybe we can wait until the python API is stablized enough, and then consider move the python modules out of ceph source tree. but personally, i don't think that hosting the ceph-mgr python modules in ceph/ceph would prevent more ceph-mgr python modules from being created. and i agree that a stable API would encourage people to create more of them. but i think it would take time to stablize the API.

**Comment by jcsp** [comment]
A few words on APIs...

Creating a stable API is *not* a case of just not changing things in PyState.cc/PyModules.cc.  That would be some effort, but not that hard.

To have a real stable API, we would have to take all the things that we currently access from mgr modules (i.e. every ::dump method on a cluster map, every JSON-formatted result from a `tell` command to a daemon, ever field in the pg_summary, the health structure, etc etc) and wrap them up in something that had a consistent set of fields (and didn't e.g. change when someone changes a ::dump method anywhere in the C++ code).  It would be something similar to what's in serializers/v2.py in the rest module.

Creating a stable API like that would be a great thing, but it's a serious amount of work that nobody is about to volunteer to do, and it's not actually necessary when we have the option of just carrying modules in our repo so that they implicitly are written to work with the same version of Ceph that they are part of.

I think there's a tendency see "python module" and assume that it's meant for out-of-tree plugins.  It's really not -- that interface exists so that people can write code in Python instead of C++.  The stability of the interface, or which tree the code lives in, is just a completely separate question from the language.  Building a stable API to Ceph is a largely separate project from writing a C++-python interface that lets us access our (unstable) calls and structures from a high level language.

**Comment by tserong** [comment]
> Should it be in ceph-mgr package or a separate package?

Speaking generally about ceph-mgr python modules (not specifically this one), depending on the module it could either be shipped in ceph-mgr, or we could do ceph-mgr-* subpackages. IMO that's a decision to make on a module by module basis. Something that was heavy with weird dependencies, I'd be inclined to have in a separate package, partly to make the dependencies clear, and partly to give distros the choice of not installing that piece. Something that was light on deps, particularly if it's something useful to most users, I'd leave in ceph-mgr.

**Comment by jcsp** [comment]
I have trimmed down the PR to just the dashboard bits.

To run this, you need to stack https://github.com/ceph/ceph/pull/14578 -> https://github.com/ceph/ceph/pull/14579 -> this

**Pull request closed without merging**